### PR TITLE
fix(#2004): remove unintentional captures in path_to_matching_str

### LINKED ISSUE #2004: Error opening directory with deep directory stucture and many underscores in the path name
### Description

opening a deep directory tree with many underscores in path name crashes with:

```
Error executing Lua callback: ...kting/.vim/plugged/nvim-tree.lua/lua/nvim-tree/utils.lua:62: too many captures
stack traceback:
        [C]: in function 'gsub'
        ...kting/.vim/plugged/nvim-tree.lua/lua/nvim-tree/utils.lua:62: in function 'path_relative'
        ...plugged/nvim-tree.lua/lua/nvim-tree/explorer/filters.lua:76: in function 'custom'
        ...plugged/nvim-tree.lua/lua/nvim-tree/explorer/filters.lua:127: in function 'should_filter'
        ...plugged/nvim-tree.lua/lua/nvim-tree/explorer/explore.lua:33: in function 'populate_children'
        ...plugged/nvim-tree.lua/lua/nvim-tree/explorer/explore.lua:68: in function 'explore'
        ...im/plugged/nvim-tree.lua/lua/nvim-tree/explorer/init.lua:28: in function '_load'
        ...im/plugged/nvim-tree.lua/lua/nvim-tree/explorer/init.lua:21: in function 'new'
        .../kting/.vim/plugged/nvim-tree.lua/lua/nvim-tree/core.lua:18: in function 'init'
        ...2/kting/.vim/plugged/nvim-tree.lua/lua/nvim-tree/lib.lua:161: in function 'open'
        /usr2/kting/.vim/plugged/nvim-tree.lua/lua/nvim-tree.lua:181: in function 'open'
```

### Neovim version

```text
NVIM v0.8.3
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
```

### Operating system and version

Ubuntu 20

### nvim-tree version

4222bb8

### Minimal config

```lua
Any config that enables nvim-tree is fine
```

### Steps to reproduce
Looks like more than 32 levels triggers it
````
1. mkdir -p _/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_
2. cd _/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_
3. touch test.txt
4. nvim
5. :NvimTreeOpen
````

### Expected behavior

Directory opens

### Actual behavior

crashes. added some prints in `path_relative`

```
relative_to: /home/kting/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_
path: /home/kting/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/_/test.txt
matching_str /home/kting/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/(%_)/
 34
final:
Error executing Lua callback: ...kting/.vim/plugged/nvim-tree.lua/lua/nvim-tree/utils.lua:66: too many captures
stack traceback:
        [C]: in function 'gsub'
        ...kting/.vim/plugged/nvim-tree.lua/lua/nvim-tree/utils.lua:66: in function 'path_relative'
        ...plugged/nvim-tree.lua/lua/nvim-tree/explorer/filters.lua:76: in function 'custom'
        ...plugged/nvim-tree.lua/lua/nvim-tree/explorer/filters.lua:127: in function 'should_filter'
        ...plugged/nvim-tree.lua/lua/nvim-tree/explorer/explore.lua:33: in function 'populate_children'
        ...plugged/nvim-tree.lua/lua/nvim-tree/explorer/explore.lua:68: in function 'explore'
        ...im/plugged/nvim-tree.lua/lua/nvim-tree/explorer/init.lua:28: in function '_load'
        ...im/plugged/nvim-tree.lua/lua/nvim-tree/explorer/init.lua:21: in function 'new'
        .../kting/.vim/plugged/nvim-tree.lua/lua/nvim-tree/core.lua:18: in function 'init'
        ...e/kting/.vim/plugged/nvim-tree.lua/lua/nvim-tree/lib.lua:161: in function 'open'
        /home/kting/.vim/plugged/nvim-tree.lua/lua/nvim-tree.lua:181: in function 'open'
        /home/kting/.vim/plugged/nvim-tree.lua/lua/nvim-tree.lua:360: in function </home/kting/.vim/plugged/nvim-tree.lua/lua/nvim-tree.lua:359>
````

### PR DESCRIPTION
**Pull request by kting28**
fixes #2004

The captures creates a limit where input path cannot have more than 32 special charactors ( `.`  , `_` or `-`)

### CODE REVIEW

File: lua/nvim-tree/utils.lua [hunk 1]
```diff
@@ -12,7 +12,7 @@ M.is_wsl = vim.fn.has "wsl" == 1
 M.is_windows = vim.fn.has "win32" == 1 or vim.fn.has "win32unix" == 1
 
 function M.path_to_matching_str(path)
-  return path:gsub("(%-)", "(%%-)"):gsub("(%.)", "(%%.)"):gsub("(%_)", "(%%_)")
+  return path:gsub("(%-)", "%%-"):gsub("(%.)", "%%."):gsub("(%_)", "%%_")
```
Comment (alex-courtis) [inline lua/nvim-tree/utils.lua:15]: That definitely fixes it. Those captures weren't necessary.

**Review by alex-courtis — CHANGES_REQUESTED** [review]
This fixes the case for underscores and dashes, however there is still potential for other magic regex characters to cause an issue.

A more complete solution might be for `path_relative` to use string.find (plain) and string.sub to extract the relative path without needing to resort to regexes.

Are you keen to make that change?

**Comment by kting28** [comment]
> to use string.find (plain)

Hmm.. looks like string.find expects a pattern so there's no dodging of striping the escape charactors. From the manual:
> The basic use of string.find is to search for a *pattern* inside a given string,

Is there a version that's straigthforward string match? I tried `path:find(...` and it requires converting into to match string first...

**Comment by alex-courtis** [comment]
> Is there a version that's straigthforward string match? I tried `path:find(...` and it requires converting into to match string first...

You can use the third argument `plain` to skip regex parsing. See https://github.com/nvim-tree/nvim-tree.lua/blob/master/lua/nvim-tree/git/init.lua#L43

**Comment by kting28** [comment]
Updated PR. I kept the original path_to_matching_str (with captures removed) for reference

**Comment by alex-courtis** [comment]
> Updated PR. I kept the original path_to_matching_str (with captures removed) for reference

No need; please remove.

File: lua/nvim-tree/utils.lua [hunk 2]
```diff
@@ -59,7 +59,15 @@ end
 ---@param relative_to string
 ---@return string
 function M.path_relative(path, relative_to)
-  local p, _ = path:gsub("^" .. M.path_to_matching_str(M.path_add_trailing(relative_to)), "")
+  local _, r = path:find(M.path_add_trailing(relative_to), 1, true)
+  local p = path
+  if r and r < p:len() then
```
Comment (alex-courtis) [inline lua/nvim-tree/utils.lua:64]: Looking good... except when there is a trailing slash.

See test case attached.
[2005-test.lua.gz](https://github.com/nvim-tree/nvim-tree.lua/files/10776248/2005-test.lua.gz)

**Review by alex-courtis — CHANGES_REQUESTED** [review]
Many thanks for going the extra mile on this one.

There are a lot of windows standard paths that contain magic characters and this fix will take care of all of them.

File: lua/nvim-tree/utils.lua (hunk 2 above)
Comment (kting28) [reply lua/nvim-tree/utils.lua]: Thanks for the test cases. Just pushed the fix.  previously version if relative == path, I kept the path but to match the original gsub version, the code now just returns ""

Comment (alex-courtis) [reply lua/nvim-tree/utils.lua]: Fantastic, thank you.

**Review by alex-courtis — APPROVED** [review]
Thank you for your contribution

File: lua/nvim-tree/utils.lua (hunk 2 above)
Comment (kting28) [reply lua/nvim-tree/utils.lua]: no problem, thanks for the excellent plugin!

**Pull request merged**